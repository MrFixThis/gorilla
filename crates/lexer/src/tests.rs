use crate::TokenKind::Eof;
use std::collections::VecDeque;

use super::*;

macro_rules! snapshot {
    ( $( ($func:ident, $path:expr) ),* $(,)? ) => {
        $(
            #[test]
            fn $func() {
                let src = include_str!($path);
                let mut conf = ::insta::Settings::clone_current();
                conf.set_snapshot_path("../testdata/output/");
                conf.bind(|| ::insta::assert_snapshot!(prepare_snapshot(src)));
            }
        )*
    };
}

fn prepare_snapshot(src: &str) -> String {
    let lexer = Lexer::new(src);
    let mut toks: VecDeque<_> = lexer.collect_tokens();
    let mut out = String::new();

    for (line, row) in src.lines().enumerate() {
        out.push_str(row);
        out.push('\n');

        while let Some(tok) = toks.pop_front() {
            if tok.span.start_line != line && tok.kind != Eof {
                toks.push_front(tok);
                break;
            }

            out.push_str(&" ".repeat(tok.span.start_col));
            out.push_str(&"^".repeat((tok.span.end_col - tok.span.start_col) + 1));
            out.push_str(&format!(" {tok:?}\n"));
        }
    }

    out
}

snapshot!(
    (
        test_code_snipet_tokenization,
        "../testdata/snapshots/code_snipet.gl"
    ),
    (
        test_comment_tokenization,
        "../testdata/snapshots/comment.gl"
    ),
    (
        test_container_types_tokenization,
        "../testdata/snapshots/container_type.gl"
    ),
    (
        test_literal_error_tokenization,
        "../testdata/snapshots/literal_error.gl"
    ),
    (
        test_function_tokenization,
        "../testdata/snapshots/function.gl"
    ),
    (
        test_identifier_tokenization,
        "../testdata/snapshots/identifier.gl"
    ),
    (
        test_literal_tokenization,
        "../testdata/snapshots/literal.gl"
    ),
);
