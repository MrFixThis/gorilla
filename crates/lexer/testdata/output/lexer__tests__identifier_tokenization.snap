---
source: crates/lexer/src/tests.rs
expression: prepare_snapshot(src)
---
pub func just_a_func() {
^^^ Keyword(Pub) -> [0 - 0, 0 - 2]
    ^^^^ Keyword(Func) -> [0 - 4, 0 - 7]
         ^^^^^^^^^^^ Ident(just_a_func) -> [0 - 9, 0 - 19]
                    ^ OpenDel(Parenthesis) -> [0 - 20, 0 - 20]
                     ^ CloseDel(Parenthesis) -> [0 - 21, 0 - 21]
                       ^ OpenDel(Brace) -> [0 - 23, 0 - 23]
    var a = "gorilla";
    ^^^ Keyword(Var) -> [1 - 4, 1 - 6]
        ^ Ident(a) -> [1 - 8, 1 - 8]
          ^ Eq -> [1 - 10, 1 - 10]
            ^^^^^^^^ Literal(Str, "gorilla") -> [1 - 12, 1 - 19]
                     ^ Semicolon -> [1 - 21, 1 - 21]
    const = a += " lang!";
    ^^^^^ Keyword(Const) -> [2 - 4, 2 - 8]
          ^ Eq -> [2 - 10, 2 - 10]
            ^ Ident(a) -> [2 - 12, 2 - 12]
              ^^ BinOpEq(Plus) -> [2 - 14, 2 - 15]
                 ^^^^^^^ Literal(Str, " lang!") -> [2 - 17, 2 - 23]
                         ^ Semicolon -> [2 - 25, 2 - 25]

    println(a)
    ^^^^^^^ Ident(println) -> [4 - 4, 4 - 10]
           ^ OpenDel(Parenthesis) -> [4 - 11, 4 - 11]
            ^ Ident(a) -> [4 - 12, 4 - 12]
             ^ CloseDel(Parenthesis) -> [4 - 13, 4 - 13]
}
^ CloseDel(Brace) -> [5 - 0, 5 - 0]
^ Eof -> [5 - 0, 5 - 0]
